sprout/appliances/tasks.py

```diff
@@ -152,6 +152,130 @@ def kill_unused_appliances(self):
+                    appliance.delete()
+                else:
+                    raise  # No diaper pattern here!
+    # If something happened to the appliance provisioning process, just delete it to remove
+    # the garbage. It will be respinned again by shepherd.
+    # Grace time is specified in BROKEN_APPLIANCE_GRACE_TIME
+    expiration_time = (timezone.now() - timedelta(**settings.BROKEN_APPLIANCE_GRACE_TIME))
+    for appliance in Appliance.objects.filter(ready=False, marked_for_deletion=False).all():
+        if appliance.status_changed < expiration_time:
+            self.logger.info("Killing broken appliance {}/{} "
+                             "on provider {}".format(appliance.id, appliance.name,
+                                                     appliance.provider.id))
+            Appliance.kill(appliance)  # Use kill because the appliance may still exist
+    # And now - if something happened during appliance deletion, call kill again
+    for appliance in Appliance.objects.filter(
+            marked_for_deletion=True, status_changed__lt=expiration_time).all():
+        self.logger.info(
+            "Trying to kill unkilled appliance {}/{} "
+            "on provider {}".format(appliance.id, appliance.name, appliance.provider.id))
+        Appliance.kill(appliance, force_delete=True)
+
+
+@singleton_task()
+def kill_lost_appliances(self):
+    """Looks for lost appliances present on sprout provider and kills them"""
+    self.logger.info("looking for old lost vms on sprout providers")
+    for provider in Provider.objects.filter(disabled=False, working=True, appliance_limit__gt=0):
+        kill_lost_appliances_per_provider(provider.id)
+
+
+@singleton_task()
+def kill_lost_appliances_per_provider(self, provider_id):
+    """Looks for lost appliance in some certain provider"""
+    rules = settings.CLEANUP_RULES
+    provider = Provider.objects.get(id=provider_id)
+    self.logger.info("obtaining list of vms on provider {}".format(provider.id))
+    try:
+        vms = provider.api.list_vms()
+        vm_names = vms if provider.provider_type == 'openshift' else [vm.name for vm in vms]
+        # skipping appliances present in sprout db. those will be handled by another task
+        vm_names = [name for name in vm_names
+                    if not Appliance.objects.filter(name=name, template__provider=provider)]
+        # checking vm time
+        for rule in rules:
+            expiration_time = timezone.now() - timedelta(**rule['lifetime'])
+            for name in vm_names:
+                if not re.match(rule['name'], name):
+                    continue
+
+                if provider.provider_type != 'openshift':
+                    try:
+                        vm_creation_time = provider.api.get_vm(name).creation_time
+                    except Exception as e:
+                        self.logger.exception("Couldn't get vm {vm} timestamp (prov: {p}) "
```

Comment:
If this exception occurs, vm_creation_time won't be set and L240 will raise exception yes?

Reply:
it will just put exception warn in the log file and proceed to next vm

Reply:
Nvm, the continue breaks us out.